xprof user guide (#2300)
AI-Hypercomputer/maxtext: docs/guides/xprof_user_guide.md

Author: Updated the doc with this linked now.

Reviewer: can you clarify if you are making this update in a separate PR or if it's not needed for this update?

# XProf for MaxText developers

## Introduction to XProf

> You can use XProf to profile and analyze the training performance of AI models. XProf helps you understand how to optimize model performance, identify bottlenecks, and improve training efficiency.

Reviewer 2: ```suggestion
[XProf](https://openxla.org/xprof) is a profiling and performance analysis tool for machine learning.

You can use XProf to profile and analyze the training performance of AI models. XProf helps you understand how to optimize model performance, identify bottlenecks, and improve training efficiency.
```

## Profiling in JAX 

XProf supports profiling JAX models, which is crucial for MaxText developers working with JAX. You can profile your JAX models using various methods, including:

*   **Sampling Mode** This mode allows for continuous profiling by sampling data during model execution.
*   **Programmatic Mode:** This provides more granular control over when and what to profile, allowing you to instrument your code with specific profiling markers. This method is integrated with MaxText code.

> Below is an example of how programmatic XProf works with Python code.
> 
> 
> ```
> JAX.profiler.trace(profile_diretory)
> batch = next(data_iterator)
> with JAX.profiler.StepTraceAnnotation("train_step", step_num=i):
>   state, metrics = train_step(state, batch)
> JAX.profiler.stop()
> ```

Reviewer 2: What are the requirements for a user to run this code? Should they install maxtext and xprof? Can we give folks a complete example, maybe using simple data just to verify how this works?

```suggestion
Below is an example of how programmatic XProf works with Python code.

```python
import jax

jax.profiler.trace(profile_diretory)  # 
batch = next(data_iterator)
with jax.profiler.StepTraceAnnotation("train_step", step_num=i):
  state, metrics = train_step(state, batch)
jax.profiler.stop()
```
```

> `JAX.profiler.TraceAnnotation` is a tool for adding custom annotations to JAX traces.

Reviewer 2: ```suggestion
[`jax.profiler.TraceAnnotation`](https://docs.jax.dev/en/latest/_autosummary/jax.profiler.TraceAnnotation.html) is a tool for adding custom annotations to JAX traces. 
```

Can we clarify how this connects with the example above?

> ## For more detailed tool illustration see 
> [XProf Detained Feature Document from OpenXLA](https://openxla.org/XProf)

Reviewer 2: ```suggestion
For more details, including tool illustrations, see 
[XProf Detained Feature Document from OpenXLA](https://openxla.org/XProf).
```

# XProf for MaxText developers

## Introduction to XProf

[XProf](https://openxla.org/xprof) is a profiling and performance analysis tool for machine learning.

You can use XProf to profile and analyze the training performance of AI models. XProf helps you understand how to optimize model performance, identify bottlenecks, and improve training efficiency.

## Profiling in JAX 

XProf supports profiling JAX models, which is crucial for MaxText developers working with JAX. You can profile your JAX models using various methods, including:

*   **Sampling Mode** This mode allows for continuous profiling by sampling data during model execution.
*   **Programmatic Mode:** This provides more granular control over when and what to profile, allowing you to instrument your code with specific profiling markers. This method is integrated with MaxText code.

> Below is an example of how programmatic XProf works with Python code.

Reviewer: How about:  The following example shows how to trace a JAX operation in Python.

```
import jax
import jax.numpy as jnp
import jax.profiler

> profile_diretory='dir_to_upload_profile'

Reviewer: typo in `directory`

jax.profiler.trace(profile_diretory)
x = jnp.ones((1000, 1000))
iter = 0
with jax.profiler.StepTraceAnnotation("dot_product", step_num=iter):
    result = jnp.dot(x, x.T).block_until_ready()
    iter += 1
jax.profiler.stop()
```

[`jax.profiler.TraceAnnotation`](https://docs.jax.dev/en/latest/_autosummary/jax.profiler.TraceAnnotation.html) is a tool for adding custom annotations to JAX traces. 

For more details, including tool illustrations, see 

> [XProf Detained Feature Document from OpenXLA](https://openxla.org/XProf).

Reviewer: - link is broken; use lowercase `xprof` in the URL
- `detained`-->`detailed`. 
- Recommend rewriting to: For more information about XProf tools, see the [XProf documentation](https://openxla.org/xprof).

The following example shows how to trace a JAX operation in Python.

```
import jax
import jax.numpy as jnp
import jax.profiler

profile_directory='dir_to_upload_profile'
jax.profiler.trace(profile_directory)
x = jnp.ones((1000, 1000))
iter = 0
with jax.profiler.StepTraceAnnotation("dot_product", step_num=iter):
    result = jnp.dot(x, x.T).block_until_ready()
    iter += 1
jax.profiler.stop()
```

You can use [`jax.profiler.TraceAnnotation`](https://docs.jax.dev/en/latest/_autosummary/jax.profiler.TraceAnnotation.html) to add custom annotations to JAX traces.

## Profiling configuration in MaxText

The following parameters control how profiling is executed within MaxText, allowing you to capture detailed performance data for analysis.

`profiler`	specifies the profiler backend to use for capturing performance traces.	Options can be `xplane`, `nsys`. Default is "".	`xplane` is for XLA/TPU and `nsys` is for CUDA/GPU.

`profiler_steps` defines the total number of steps to run during the profiling capture window.	Default is 5

> `skip_first_n_steps_for_profiler` specifies the number of initial training steps to skip before the profiling capture begins. This is typically used to bypass model warmup and capture steady-state performance.	default is 1.

Reviewer 3: Can you make these a bulleted list?

> # Xprof User Guide for Maxtext Developers

Author: updated

## Introduction to Xprof

> Xprof is a powerful tool designed for profiling and analyzing the training performance of AI models. For Maxtext developers, understanding and utilizing Xprof can significantly help in optimizing model performance, identifying bottlenecks, and improving training efficiency.

Author: This xprof is using jax profiler and is accessible by cloud env. We will keep it now and change to [cloud version](https://github.com/AI-Hypercomputer/cloud-diagnostics-xprof) when we have alignment.

> Xprof is a powerful tool designed for profiling and analyzing the training performance of AI models. For Maxtext developers, understanding and utilizing Xprof can significantly help in optimizing model performance, identifying bottlenecks, and improving training efficiency.

Author: Thanks, good suggestions!

## Profiling Jax 

Xprof supports profiling Jax models, which is crucial for Maxtext developers working with Jax. You can profile your Jax models using various methods, including:

> *   **Sampling Mode:** This mode allows for continuous profiling by sampling data during model execution.

Reviewer 3: Continuin the offline discussion with @suexu1025 and @RissyRan. Since MaxText doesn't support sampling mode, can you move it after programmatic profiling? And also add a sentence to this saying this mode is not implemented in MaxText out of the box with a link to the doc you just shared here

*   **Programmatic Mode:** This provides more granular control over when and what to profile, allowing you to instrument your code with specific profiling markers. This method is integrated with Maxtext code.

Below is an example of how programmatic Xprof works with Python code.

```
jax.profiler.trace(profile_diretory)
batch = next(data_iterator)
with jax.profiler.StepTraceAnnotation("train_step", step_num=i):
  state, metrics = train_step(state, batch)
jax.profiler.stop()
```

`jax.profiler.TraceAnnotation` is a tool for adding custom annotations to JAX traces.

## Xprof UI Features

The Xprof User Interface (UI) offers several key features to help you analyze your profiling data:

> *   Trace Viewer

Author: I changed the doc to link openxla doc

# XProf for MaxText developers

## Introduction to XProf

You can use XProf to profile and analyze the training performance of AI models. XProf helps you understand how to optimize model performance, identify bottlenecks, and improve training efficiency.

> ## Profiling in JAX 

Reviewer 3: Could you add the MaxText flags that are used for profiling here? E.g. `profiler=xplane`, `profiler_steps`, `skip_first_n_steps_for_profiler`, etc.?

Author: Added configuration details.

```

You can use [`jax.profiler.TraceAnnotation`](https://docs.jax.dev/en/latest/_autosummary/jax.profiler.TraceAnnotation.html) to add custom annotations to JAX traces.

> For more information about XProf tools, see the [XProf documentation](https://openxla.org/XProf).

Reviewer: still need to correct the capitalization in the URL to `xprof`
